### DISCUSSION TITLE
How to Bundle Dynamic Libraries and Their Resources

### DISCUSSION THREAD

**Discussion by WSH032** [tauri-apps/tauri#11883] [Q&A]
Hi Tauri developers, I am working on integrating pytauri with tauri-cli (WSH032/pytauri#2). To achieve this, I need to embed the Python interpreter into the executable, which requires the executable to link with the Python dynamic library. Therefore, I need to bundle and distribute a [portable Python](https://github.com/indygreg/python-build-standalone).

Here are the Python dependencies I bundled and their file layouts:

- [Windows](https://github.com/indygreg/python-build-standalone/releases/download/20241016/cpython-3.10.15+20241016-x86_64-pc-windows-msvc-install_only.tar.gz):

    ```tree
    ├── python3.dll
    ├── python.exe
    ├── ...
    ```

- [Linux](https://github.com/indygreg/python-build-standalone/releases/download/20241016/cpython-3.10.15+20241016-x86_64-unknown-linux-gnu-install_only.tar.gz):

    ```tree
    ├── bin
    │   ├── python
    │   └── ...
    ├── lib
    │   ├── libpython3.so
    │   └── ...
    ├── share
    │   └── ...
    └── include
        └── ...
    ```

- [Apple](https://github.com/indygreg/python-build-standalone/releases/download/20241016/cpython-3.10.15+20241016-x86_64-apple-darwin-install_only.tar.gz): Layout is similar to Linux

---

For Windows, it's easy. Just place the executable `pytauri-demo.exe` in the root directory, like this:

```tree
├── python3.dll
├── python.exe
├── pytauri-demo.exe  <--
```

It seems that I can use [bundleresources](https://v2.tauri.app/reference/config/#bundleresources) to achieve this?

---

But for Unix, I need to place the executable in `/bin` and set the dynamic library search path via `rpath=$ORIGIN/../lib`:

```tree
├── bin
│   ├── python
│   ├── pytauri-demo  <--
│   └── ...
├── lib
│   ├── libpython3.so
│   └── ...
├── share
│   └── ...
└── include
    └── ...
```

I don't know how to configure tauri-cli to achieve this layout. Do you have any suggestions?

Or is there a better way than `rpath`? My Unix experience is very limited, and I know nothing about Unix packaging (rpm, appimage, deb).

It seems that AppImage can bundle dependencies? Can we use it to bundle the portable Python I mentioned?

**Comment by BjornTheProgrammer** [comment]
You have a few options here probably. 1) Use a [resource](https://v1.tauri.app/v1/guides/building/resources/) directory, in which you manually add the Python interpreter you want to use. 2) Use something like [`pyembed`](https://docs.rs/pyembed/latest/pyembed/) to build and embed a CPython instance. This has a few drawbacks, namely pretty hard to build python reproducably. 3) You could try to write bindings for [python-build-standalone](https://github.com/indygreg/python-build-standalone), again this also forces your end user to have `gcc` installed, although you can probably just make a script to download the python release from the repo. 4) You could try to use something like [`RustPython`](https://github.com/RustPython/RustPython) to embed Python and run the python scripts which boot tauri. This makes it so you don't have to install anything other than Rust. But `RustPython` is slower.

You need to have a `main.rs` function to be built into a packaged dependency, and that seems to be what you might be missing. You need to add the embedded interpreter into that.

I've created an WIP example [pull request](https://github.com/WSH032/pytauri/pull/3) for you to view that shows what I'm talking about with `RustPython`.
